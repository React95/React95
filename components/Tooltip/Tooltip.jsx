import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Tip = styled.div`
  background: radial-gradient(#ff0 20%, transparent 20%) 0 0,
    radial-gradient(#ff0 20%, transparent 20%) 4px 4px,
    radial-gradient(rgba(255, 255, 255, 0.1) 20%, transparent 25%) 0 1px,
    radial-gradient(rgba(255, 255, 255, 0.1) 20%, transparent 25%) 3px 4px;
  background-size: 7px 7px;
  background-color: #fff;
  border: 1px solid #000;
  padding: 2px 2px 2px 4px;
  position: absolute;
  top: -20px;
  text-align: center;
  z-index: 100;
  display: ${({ show }) => (show ? 'block' : 'none')};
`;

const Wrapper = styled.div`
  display: inline-block;
  position: relative;
  cursor: default;
  white-space: nowrap;
`;

const Tooltip = ({ children, text, delay }) => {
  const [show, setShow] = useState(false);
  const [delayTimer, setDelayTimer] = useState(null);

  const handleEnter = () => {
    const timer = setTimeout(() => {
      setShow(true);
    }, delay);

    setDelayTimer(timer);
  };

  const handleLeave = () => {
    clearTimeout(delayTimer);
    setShow(false);
  };

  return (
    <Wrapper onMouseEnter={handleEnter} onMouseLeave={handleLeave}>
      <Tip show={show}>{text}</Tip>
      {children}
    </Wrapper>
  );
};

function formatDate(date) {
  const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];

  const day = date.getDate();
  const monthIndex = date.getMonth();
  const year = date.getFullYear();

  return (
    `${day.toString().padStart(2, '0')} ${monthNames[monthIndex]} ${year}`
  );
}

Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  text: PropTypes.string,
  delay: PropTypes.number,
};

Tooltip.defaultProps = {
  delay: 1000,
  text: formatDate(new Date()),
};

export default Tooltip;
