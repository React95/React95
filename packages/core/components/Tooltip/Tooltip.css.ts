import { style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';
import { calc } from '@vanilla-extract/css-utils';

export const tip = style({
  background: `radial-gradient(#ff0 20%, transparent 20%) 0 0,
     radial-gradient(#ff0 20%, transparent 20%) 4px 4px,
     radial-gradient(rgba(255, 255, 255, 0.1) 20%, transparent 25%) 0 1px,
     radial-gradient(rgba(255, 255, 255, 0.1) 20%, transparent 25%) 3px 4px`,
  backgroundSize: `${contract.space[7]} ${contract.space[7]}`,
  backgroundColor: contract.colors.borderLightest,
  borderWidth: contract.space[1],
  borderStyle: 'solid',
  paddingTop: contract.space[2],
  paddingRight: contract.space[2],
  paddingBottom: contract.space[2],
  paddingLeft: contract.space[4],
  boxShadow: 'none',
  position: 'absolute',
  top: calc.negate(contract.space[20]),
  textAlign: 'center',
  zIndex: contract.zIndices.taskbar,
});

export const wrapper = style({
  display: 'inline-block',
  position: 'relative',
  cursor: 'default',
  whiteSpace: 'nowrap',
});
