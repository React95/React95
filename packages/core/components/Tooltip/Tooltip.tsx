import React, { useState, forwardRef } from 'react';
import type { HTMLAttributes, Ref } from 'react';
import cn from 'classnames';

import { tip, wrapper } from './Tooltip.css';

function formatDate(date: Date): string {
  const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];

  const day = date.getDate();
  const monthIndex = date.getMonth();
  const year = date.getFullYear();

  return `${day.toString().padStart(2, '0')} ${monthNames[monthIndex]} ${year}`;
}

export type TooltipProps = {
  text?: string;
  delay?: number;
} & HTMLAttributes<HTMLDivElement>;

const TooltipRenderer = (
  {
    children,
    text = formatDate(new Date()),
    delay = 1000,
    ...rest
  }: TooltipProps,
  ref: Ref<HTMLDivElement>,
) => {
  const [show, setShow] = useState(false);
  const [delayTimer, setDelayTimer] = useState(0);

  const handleEnter = () => {
    const timer = window.setTimeout(() => {
      setShow(true);
    }, delay);

    setDelayTimer(timer);
  };

  const handleLeave = () => {
    clearTimeout(delayTimer);
    setShow(false);
  };

  return (
    <div
      {...rest}
      className={cn(wrapper, rest.className)}
      onMouseEnter={handleEnter}
      onMouseLeave={handleLeave}
      ref={ref}
    >
      {show && <div className={tip}>{text}</div>}
      {children}
    </div>
  );
};

export const Tooltip = forwardRef<HTMLDivElement, TooltipProps>(
  TooltipRenderer,
);
