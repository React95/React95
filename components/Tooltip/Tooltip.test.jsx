import React from 'react';
import { render, fireEvent, act } from 'react-testing-library';
import Tooltip from './Tooltip';

describe('<Tooltip />', () => {
  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <Tooltip delay={0} text="Tooltip text">
          <span>hover</span>
        </Tooltip>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('Tooltip visibility', () => {
    it('should show Tooltip when hovered', () => {
      const tooltipText = 'Tooltip text';
      jest.useFakeTimers();

      const { container, getByText } = render(
        <Tooltip delay={0} text={tooltipText}>
          <span>hover</span>
        </Tooltip>
      );

      expect(getByText(tooltipText)).toHaveStyleRule('display', 'none');

      act(() => {
        fireEvent.mouseEnter(container.firstChild);
        jest.runAllTimers();
      });

      expect(getByText(tooltipText)).toHaveStyleRule('display', 'block');
    });
  });
});
