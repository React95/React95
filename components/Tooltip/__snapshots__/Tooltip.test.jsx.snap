// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> Snapshot should match snapshot 1`] = `
.c1 {
  background: radial-gradient(#ff0 20%,transparent 20%) 0 0,radial-gradient(#ff0 20%,transparent 20%) 4px 4px,radial-gradient(rgba(255,255,255,0.1) 20%,transparent 25%) 0 1px,radial-gradient(rgba(255,255,255,0.1) 20%,transparent 25%) 3px 4px;
  background-size: 7px 7px;
  background-color: #fff;
  border: 1px solid #000;
  padding: 2px 2px 2px 4px;
  position: absolute;
  top: -20px;
  text-align: center;
  z-index: 100;
  display: none;
}

.c0 {
  display: inline-block;
  position: relative;
  cursor: default;
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Tooltip text
    </div>
    <span>
      hover
    </span>
  </div>
</div>
`;
